jcstress-samples/src/main/java/org/openjdk/jcstress/samples/high/race/condition/RaceCondition_01_ReadModifyWriteSequence.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.jcstress.samples.high.race.condition;
+
+import org.openjdk.jcstress.annotations.*;
+import org.openjdk.jcstress.infra.results.III_Result;
+import org.openjdk.jcstress.infra.results.I_Result;
+
+import static org.openjdk.jcstress.annotations.Expect.*;
+
+/*
+    How to run this test:
+        $ java -jar jcstress-samples/target/jcstress.jar -t RaceCondition_01_ReadModifyWriteSequence
+ */
+
+/**
+ * This sample demonstrates you how a read-modify-write sequence can lead to surprising results.
+ */
+@JCStressTest
+@Outcome(id = {"150, 100, 150"}, expect = ACCEPTABLE, desc = "Actor1 considered actor2's result and wrote its right result")
+@Outcome(id = {"250, 150, 150"}, expect = ACCEPTABLE, desc = "Actor2 considered actor1's result and wrote its right result")
+@Outcome(id = {"250, 100, 250", "250, 150, 250"}, expect = ACCEPTABLE_INTERESTING, desc = "Actor1 ignored actor2's result and wrote its wrong result")
+@Outcome(id = {"250, 100, 100", "150, 100, 100"}, expect = ACCEPTABLE_INTERESTING, desc = "Actor2 ignored actor1's result and wrote its wrong result")
```

Comment:
Please explain how `"250, 150, 250"` and `"150, 100, 100"` are acceptable.

I think those outcomes are forbidden.

Reply:
Why should they be forbidden?
I think they are undesirable but allowed because the lack of the appropiate synchronisation causes these result within a read-modify-write-sequence.

Reply:
>Why should they be forbidden?

> 17.4.4. Synchronization Order: https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.4
> A write to a volatile variable v (§8.3.1.4) synchronizes-with all subsequent reads of v by any thread (where "subsequent" is defined according to the synchronization order).

Reply:
I think the synchronization with volatile is just too weak. It keeps the order of the statements, yes, that's right,
However, it doesn't force orders with the desired result without the effect of the read-modify-write sequence. 
We can remove volatile if you like.

Reply:
> the order of the statements

The order of statements is "program order" which always holds no matter what ("final field update" is one of the exceptions, yet updating a final field is really weird)

> it doesn't force orders

Please use jls-17 when you reason on the acceptable by Java memory model values.
See https://shipilev.net/blog/2014/jmm-pragmatics/ ,

> In order to test if an execution is plausible under JMM, you need to see if it passes all the requirements. Note, however, that you can quickly branch-and-bound the set of considered executions based on the failure of a particular test....

Reply:
I have followed your advice to use traces instead of the wordy eplaination.
You can see now the traces:
@Outcome(id = {"250, 100, 250"}, expect = ACCEPTABLE_INTERESTING, desc = "1:v=200, 1:newV=250, 2:v=200, 2:newV=100, 2:v=newV, 1:v=newV")
@Outcome(id = {"250, 100, 100"}, expect = ACCEPTABLE_INTERESTING, desc = "1:v=200, 1:newV=250, 2:v=200, 2:newV=100, 1:v=newV, 2:v=newV")
Can you tell me where in the paths they violate JMM rules? 
I would really like to understand that before I change that.
If I don't understand it finally but you insist on it, then I will change it, of course.

@shipilev: what do you think?

Reply:
Can we resolve that now, @vlsi?